lib/execution_segment.go

```diff
@@ -394,13 +396,275 @@ func (ess ExecutionSegmentSequence) String() string {
+// We use the below algorithm to split elements between ExecutionSegments by using their length as
+// the rational number. As we would like to get non sequential elements we try to get the maximum
+// distance between them. That is the number of elements divided by the number of elements for any
+// given segment, which concidently is the length of the segment reversed.
+// The algorithm below does the following:
+// 1. Goes through the elements from 0 to the lcd-1
+// 2. For each of element goes through the segments and looks if the amount of already taken
+// elements by the given segment multiplied by that segment length inverted is equal to or less to
+// the current element index. if it is give that element to that segment if not continue with the
+// next element.
 //
-// TODO: add a more detailed algorithm description
-func (ess ExecutionSegmentSequence) GetStripedOffsets(segment *ExecutionSegment) (int, []int, error) {
-	start := 0
-	offsets := []int{}
-	// TODO: basically https://docs.google.com/spreadsheets/d/1V_ivN2xuaMJIgOf1HkpOw1ex8QOhxp960itGGiRrNzo/edit
-	return start, offsets, fmt.Errorf("not implemented")
+// The code below specifically avoids using big.Rat which complicates the code somewhat.
+// As additional note the sorting of the segments from biggest to smallest helps with the fact that
+// the biggest elements will need to take the most elements and for them it will be the hardest to
+// not get sequential elements.
+func (e sortInterfaceWrapper) stripingAlgorithm(saveIndex func(iteration int64, index int, numerator int64) bool) {
+	var chosenCounts = make([]int64, len(e.slice))
+
+outer:
+	for i := int64(0); i < e.lcd; i++ {
+		for index, chosenCount := range chosenCounts {
+			num := chosenCount * e.lcd
+			denom := e.slice[index].numerator
+			if i > num/denom || (i == num/denom && num%denom == 0) {
+				chosenCounts[index]++
+				if saveIndex(i, e.slice[index].originalIndex, denom) {
+					break outer
+				}
+				break
+			}
+		}
+	}
+}
+
+// ExecutionTuple is here to represent the combination of ExecutionSegmentSequence and
+// ExecutionSegment and to give easy access to a couple of algorithms based on them in a way that is
+// somewhat perfomant for which it generally needs to cache the results
+type ExecutionTuple struct { // TODO rename
+	ES *ExecutionSegment // TODO unexport this as well?
+
+	// TODO: have the index of the segment, cached?
+	sequence     ExecutionSegmentSequence
+	offsetsCache [][]int64
+	lcd          int64
+	// TODO discuss if we just don't want to fillCache in the constructor and not need to use pointer receivers everywhere
+	once *sync.Once
+}
+
+func fillSequence(sequence ExecutionSegmentSequence) ExecutionSegmentSequence {
+	// TODO: discuss if we want to get the lcd of the sequence and fill with it elements of length 1/lcd ?
+	if sequence[0].from.Cmp(zeroRat) != 0 {
+		es, err := NewExecutionSegment(zeroRat, sequence[0].from)
+		if err != nil {
+			panic(err) // this really can't happen
```

Comment:
Hm, might just be better to ignore the error as I do below in NewExecutionTuple. So that way when you ignore the error it is obvious instead of knowing that the unexported variant is without the checks. Not to speak that the autocomplete might complete the wrong one and the person to not notice.

Reply:
You have case-insensitive autocomplate :open_mouth: Joking aside, we can name the internal function `createExecutionTuple` or something else. But considering that we don't care about its validations, it will also speed things up slightly to avoid them.

Reply:
I was going to argue the performance benefit won't be great (and I will still argue that) but it does save 3 big.Rat comparisons which is probably not so negligible

Reply:
yeah, any performance gains would be negligible and would just be a side-benefit, my biggest issue is "Go Purity" :rofl: My brain doesn't like seeing `panic()` or `_ = whatever()` :blush:

Reply:
resolved in https://github.com/loadimpact/k6/pull/1323/commits/73e6f5b403369f1e72be4129b902158bfa292f1f